packages/monaco-kubernetes/src/kubernetes.worker.ts

```diff
@@ -0,0 +1,209 @@
+  Position,
+  Range,
+  TextEdit,
+} from "vscode-languageserver-types";
+import {
+  CustomFormatterOptions,
+  getLanguageService,
+} from "yaml-language-server/lib/esm/languageservice/yamlLanguageService.js";
+
+import { languageId } from "./constants.js";
+import type { LanguageSettings } from "./index.js";
+import { MonokleService } from "./validation/validationService.js";
+
+const DEFAULT_SCHEMA_VERSION = "1.26.0";
+
+export type CreateData = LanguageSettings;
+
+export interface KubernetesWorker {
+  reconfigure(newSettings: LanguageSettings): void;
+
+  doValidation: (uri: string) => Diagnostic[];
+
+  doComplete: (uri: string, position: Position) => CompletionList;
+
+  doDefinition: (uri: string, position: Position) => LocationLink[];
+
+  doHover: (uri: string, position: Position) => Hover;
+
+  format: (uri: string, options: CustomFormatterOptions) => TextEdit[];
+
+  resetSchema: (uri: string) => boolean;
+
+  findDocumentSymbols: (uri: string) => DocumentSymbol[];
+
+  findLinks: (uri: string) => DocumentLink[];
+
+  getCodeAction: (
+    uri: string,
+    range: Range,
+    diagnostics: Diagnostic[]
+  ) => CodeAction[];
+}
+
+initialize<KubernetesWorker, CreateData>((ctx, settings) => {
+  const parser = new ResourceParser();
+  const loader = new SchemaLoader();
+  const validator = new MonokleService(parser, loader);
+
+  validator.configure(settings);
+
+  const languageService = getLanguageService(
+    async (uri: string): Promise<string> => {
+      const resource = validator.getResource(uri);
+
+      if (!resource) {
+        const response = await fetch(uri);
+        if (!response.ok) {
+          throw new Error(`Schema request failed for ${uri}`);
+        }
+        return response.text();
+      }
+
+      const version =
+        settings.validation.settings["kubernetes-schema"]?.["schemaVersion"] ??
+        DEFAULT_SCHEMA_VERSION;
+
+      const { schema } = await loader.getResourceSchema(version, resource);
+      const schemaString = JSON.stringify(schema);
+      return schemaString;
+    },
+    null,
+    null,
+    null,
+    null
+  );
+
+  languageService.configure({
+    isKubernetes: true,
+    format: true,
+    completion: true,
+    hover: true,
+    validate: false,
+    yamlVersion: "1.2",
+    customTags: [],
+    schemas: [],
+  });
+
+  languageService.registerCustomSchemaProvider((uri) => {
+    // TODO: support multi-resource YAML models.
+    // Currently, this only works for single-resource YAML models where
+    // the uri maps exactly to a one resource. To support multi-
+    // resource files we should parse the model here to generate
+    // a URL that identifies all the resource kinds. The schema
+    // request service can then interpret the URL to fetch and stitch a proper
+    // oneOf schema to validate. This requires a refactor of the SchemaLoader.
+    // @example kubernetes://io.k8s.api.apps.v1.Deployment+io.argoproj.v1alpha1.Application
+    return Promise.resolve(uri);
```

Comment:
Not sure I should keep this TODO. I'm a bit sad I didn't have time to work this out but it's one of those features that needs small cascading changes all over our libraries.

Reply:
Let's keep it - no probs with pushing this out as the initial use-case (Argo) is single-resource anyway..